server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java

```diff
@@ -539,10 +539,26 @@ private static IndexMetadata.Builder anIndex(String name, Settings.Builder setti
     private static ClusterState stateWithStartedIndices(IndexMetadata.Builder... indices) {
         var metadataBuilder = Metadata.builder();
         var routingTableBuilder = RoutingTable.builder(TestShardRoutingRoleStrategies.DEFAULT_ROLE_ONLY);
-        for (var index : indices) {
-            var build = index.build();
-            metadataBuilder.put(build, false);
-            routingTableBuilder.addAsNew(build);
+        if (randomBoolean()) {
+            for (var index : indices) {
+                var build = index.build();
+                metadataBuilder.put(build, false);
+                routingTableBuilder.addAsNew(build);
+            }
+        } else {
+            for (var index : indices) {
```

Comment:
I don't really understand how the test changes relate to the production code change here. Could you explain? Maybe add a comment to this test utility explaining the difference between the two branches?

Reply:
Sure,
Prior to this change we only tested a case when allocating from scratch (eg `balancer.allocateUnassigned()`). When doing so we pick a node with minimum weigh to allocate each new unassigned shard.

With this change we sometimes start from obviously unbalanced cluster (all shards are allocated on the same node). Here we rely on `balancer.balance()` to move shards properly. It might fail to do so if inside it generates unnecessarily high `localThreshold` that depends on `sorter.minWeightDelta() -> function.minWeightDelta(balancer, index) -> balancer.maxShardSizeBytes(index)` what this change is fixing.